src/gpt2/model.jl

```diff
@@ -0,0 +1,194 @@
+    if inputsize % num_heads != 0
+        throw(ArgumentError("`inputsize` not divisible by `num_heads`"))
+    end
+    GPT2Model(inputsize, num_heads, div(inputsize, num_heads), ffhiddensize, num_layers;
+              activation=activation,
+              p_dropout_ff=p_dropout_ff, p_dropout_attn=p_dropout_attn)
+end
+
+function GPT2Model(inputsize::Integer, num_heads::Integer, attnhiddensize::Integer,
+                   ffhiddensize::Integer, num_layers::Integer;
+                   activation=Flux.gelu, p_dropout_ff=0.1f0, p_dropout_attn=0.1f0)
+    GPT2Model(
+        Flux.Dropout(p_dropout_ff),
+        Transformers.Stacks.Stack(
+            Transformers.Stacks.@nntopo_str("x':x => $num_layers"),
+            [
+                GPT2Block(inputsize, num_heads, attnhiddensize, ffhiddensize;
+                          look_ahead=false, activation=activation,
+                          p_dropout_attn=p_dropout_attn)
+                for _ in 1:num_layers
+            ]...
+        ),
+        Flux.LayerNorm(inputsize)
+    )
+end
+
+function (gpt::GPT2Model)(input, mask=nothing; return_all_outputs::Bool=false)
+    input = gpt.dropout(input)
+    output, all_outputs = gpt.blocks(input)
+    output = gpt.final_norm(output)
+    isnothing(mask) || (output .*= mask)
+    if return_all_outputs
+        return output, all_outputs
+    else
+        return output  # size(output) == (inputsize, seq_len, batch_len)
+    end
+end
+
+
+# Predictor obtained by combining an embedding, a GPT2Model and an output layer
+
+struct GPT2Predictor{E<:Transformers.Basic.PositionEmbedding,
+                     G<:Transformers.Basic.AbstractTransformer, O}
+    embedding::E
+    gpt::G
+    output::O
+end
+
+Flux.@functor GPT2Predictor
+
+function GPT2Predictor(inputsize::Integer, outputsize::Integer, num_heads::Integer,
+                       attnhiddensize::Integer, ffhiddensize::Integer, num_layers::Integer,
+                       maxinputsize::Integer; activation=Flux.gelu,
+                       p_dropout_ff=0.1f0, p_dropout_attn=0.1f0,
+                       output_activation=identity, init_std=0.02)
+    model = GPT2Predictor(
+        Transformers.Basic.PositionEmbedding(inputsize, maxinputsize; trainable=true),
+        GPT2Model(
+            inputsize, num_heads, attnhiddensize, ffhiddensize, num_layers;
+            activation=Flux.gelu, p_dropout_ff=p_dropout_ff, p_dropout_attn=p_dropout_attn
+        ),
+        Transformers.Basic.Positionwise(Flux.Dense(inputsize, outputsize, output_activation))
+    )
+    init_weights!(model, init_std)
+end
+
+function (gpt::GPT2Predictor)(input, mask=nothing;
+                              return_all_outputs::Bool=false)
+    input = input .+ gpt.embedding(input)
+    output = gpt.gpt(input, mask, return_all_outputs=return_all_outputs)
+    return gpt.output(output)
+end
+
+
+# Layer initialization
+
+init_weights!(layer::GPT2Predictor, std=0.02)       = init_field_weights!(layer, std)
```

Comment:
Why do we need these init functions?

Reply:
The [original](https://github.com/openai/gpt-2/blob/0574c5708b094bfa0b0f6dfe3fd284d9a045acd9/src/model.py#L50-L56) source [code](https://github.com/openai/gpt-2/blob/0574c5708b094bfa0b0f6dfe3fd284d9a045acd9/src/model.py#L28-L38) as well as [HuggingFace](https://github.com/huggingface/transformers/blob/ea2600bd5f1d36f2fb61958be21db5b901e33884/src/transformers/configuration_gpt2.py#L131) use initialization functions with (mostly) the given default values.

Now that I looked over the code again, perhaps the `PositionEmbedding` [should have](https://github.com/openai/gpt-2/blob/0574c5708b094bfa0b0f6dfe3fd284d9a045acd9/src/model.py#L152-L153) 0.01 as initalization stderr.

Once again, sorry about no documentation yet; I'll fix it as soon as possible.

Reply:
I mean, these functions shouldn't be in this module or even this package. They are the util functions of Flux. If we don't have any special usage that require these, then we probably don't want them.